[pkg/ottl] add User Agent parsing

pchila opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Added a new ottl converter `UserAgent`: it parses an input string and matches against a [set of known UA regexes](https://github.com/ua-parser/uap-core/blob/master/regexes.yaml) to correctly identify user agent and its version
**Link to tracking Issue:** #32434 

**Testing:** Unit tests, E2E tests

**Documentation:** <Describe the documentation added.> Added UserAgent description in `pkg/ottl/ottlfuncs/README.md`

pchila committed:
Initial naive impl for user-agent parsing

pchila committed:
add uap-go dependency for user-agent parsing

pchila committed:
Initial implementation of user-agent parsing

pchila committed:
wire up UserAgentFactory and add e2e test

pchila committed:
add documentation for UserAgent converter

ycombinator commented:
Hi @atoulme, @TylerHelmuth, @bogdandrutu, or @evan-bradley, this PR is failing CI on the `license/snyk` and `security/snyk` checks.  I'm unable to view the details for either.  I tried to auth into Snyk using GitHub but it says I'm not a member of the organization.  Just a heads up that I've requested access. Thanks.

ycombinator commented:
/label -Stale

TylerHelmuth commented:
@pchila please handle the merge conflicts and merge in the latest from main

pchila committed:
Merge remote-tracking branch 'upstream/main' into user-agent-parsing

pchila commented:
> @pchila please handle the merge conflicts and merge in the latest from main
Done and pushed

TylerHelmuth commented on pkg/ottl/ottlfuncs/README.md:
```diff
@@ -1547,6 +1548,31 @@ Examples:
+
+`UserAgent(value)`
+
+The `UserAgent` Converter parses the string argument trying to match it against well-known user-agent strings as defined
+[here](https://github.com/ua-parser/uap-core/blob/master/regexes.yaml)
+
+The results of the parsing are returned as a map containing `user_agent.name`, `user_agent.version` and `user_agent.original`
+as defined in semconv v1.25.0.
```
```suggestion
The `UserAgent` Converter parses the string argument trying to match it against well-known user-agent strings.

`value` is a string or a path to a string.  If `value` is not a string an error is returned.

The results of the parsing are returned as a map containing `user_agent.name`, `user_agent.version` and `user_agent.original`
as defined in semconv v1.25.0.

Parsing is done using the [uap-go package](github.com/ua-parser/uap-go). The specific formats it recognizes can be found [here](https://github.com/ua-parser/uap-core/blob/master/regexes.yaml).
```

TylerHelmuth commented on pkg/ottl/ottlfuncs/func_useragent.go:
```diff
@@ -0,0 +1,46 @@
+		if err != nil {
+			return nil, err
+		}
+		// TODO: maybe we want to manage the regex definitions separately?
```
Unless the otel semconv defines a specific set, I am good with relying on the dependency to supply the regexes.

TylerHelmuth commented on pkg/ottl/ottlfuncs/func_useragent.go:
```diff
@@ -0,0 +1,46 @@
+			return nil, err
+		}
+		// TODO: maybe we want to manage the regex definitions separately?
+		parser := uaparser.NewFromSaved()
```
Do we need a new parser for each pass through the function?  Could the creation of the parser be moved to above the return function so it can be reused?

pchila replied:
When writing the function I wasn't sure what would be the lifetime of the parser compared to the evaluation context or the collector lifetimes.

As a first implementation I went for the safe route of initializing a new parser at each invocation (with the obvious performance penalty); since we are not allowing the user to pass UA string definitions/regexes at the moment we can move the parser to a package var with a `sync.Once.Do()` to initialize it lazily at the first UA parsing and reuse that parser for any following invocation of `UserAgent(...)` function

TylerHelmuth replied:
Since we not providing our own regexes yet, I think the second option is best.  Inside `userAgent`, before the returned `ottl.ExprFunc`, we can initialize the parser.  This is the strategy we use for regex.

pchila replied:
If I understood correctly option 2 should look like this ---> 095d8e44ec90f488b1436d63a9cd6d5120e3a089

TylerHelmuth commented on pkg/ottl/ottlfuncs/func_useragent_test.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+func TestUserAgentParser(t *testing.T) {
+	testCases := []struct {
```
Can you add a test case for the Collector's user agent string

TylerHelmuth replied:
https://github.com/open-telemetry/opentelemetry-collector/blob/7638bb298172330241375ed5dc3c50637caa645f/exporter/otlphttpexporter/otlp.go#L67

pchila replied:
Added in commit 7538cf0ecbb006add37328f1e01240c79a463028

Without any modifications the uap-go library parses it as "Other" :sweat_smile: 
If we want to support something specific we can either open a PR in `uap-core` repo (preferred since we can keep using the upstream dependency) or use our own regex file where we add the otel collector regex on top of the existing ones (this would mean we have to sync modifications coming from upstream into out own version of the regular expressions file)

TylerHelmuth replied:
If the dep is providing expected user agent strings defined by some standard then I think it is right and we should work to update the collector's user agent format separately.

TylerHelmuth commented on pkg/ottl/go.mod:
```diff
@@ -12,6 +12,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal v0.106.1
 	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest v0.106.1
 	github.com/stretchr/testify v1.9.0
+        github.com/ua-parser/uap-go v0.0.0-20240611065828-3a4781585db6
```
I want to make sure we use the best possible dependency for this.  How did we land on this package vs another? I see that it was updated recently so thats good.

pchila replied:
I had a look at uap-go library after reading https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32434#issuecomment-2069622134

The [uap-core](https://github.com/ua-parser/uap-core) lib already contained the set of UA regexes that we could reuse and then uap-go just contains the go code needed to parse the [regex file](https://github.com/ua-parser/uap-core/blob/master/regexes.yaml) and provide functions for matching strings against those while already providing some LRU caching for performance.

As far as which dependency we use we can always swap it for something else until we expose some of the internals (the regex file format for example) to users.

From a quick search there are a few libraries that could provide the functionality:
- https://github.com/mileusna/useragent (no caching, matching done inside big `switch` statements matching on parsed tokens)
- https://github.com/mssola/user_agent (archived by owner on Feb 2023)
- https://github.com/xojoc/useragent (pretty old library, matching based on hardcoded strings in .go files)

Given the alternatives, using a library that matches UA strings against regexes for well-known user agents didn't look like a bad starting point.

If we decide we don't want to use a third party lib we can always copy/extend/create the base regex file and/or implement our own matching and caching depending on the needs.

TylerHelmuth replied:
> As far as which dependency we use we can always swap it for something else until we expose some of the internals (the regex file format for example) to users.

Good point. I think sticking with this dep is good to start.

pchila committed:
Move User-Agent parser initialization out of the `UserAgent` function

Initialize a `uaparser.Parser` instance at the first UserAgent call and
reuse it instead of creating a new parser for every invocation.

pchila committed:
Add unit test for parsing collector UserAgent string

pchila committed:
Update pkg/ottl/ottlfuncs/README.md

Co-authored-by: Tyler Helmuth <12352919+TylerHelmuth@users.noreply.github.com>

pchila committed:
Merge remote-tracking branch 'upstream/main' into user-agent-parsing

pchila committed:
fixup! Merge remote-tracking branch 'upstream/main' into user-agent-parsing

pchila committed:
Move User-Agent parser initialization out of the user agent converter

This reverts commit 6ac23c0bd05ac1f453ee1afaead5eca7fc12361d. going for
a simpler solution where the useragent parser is instantiated just
before returning the anonymous ottl.ExprFunc[K] that uses it.
This will prevent re-instantating the parser at each `UserAgent(...)`
invocation.

TylerHelmuth commented:
@pchila we just had a release so there is a new merge conflict to handle.

pchila committed:
Merge remote-tracking branch 'upstream/main' into user-agent-parsing

pchila commented:
> @pchila we just had a release so there is a new merge conflict to handle.

@TylerHelmuth merged main into the branch and repushed

pchila committed:
Fix otelcontribcol build

pchila committed:
fixup! add documentation for UserAgent converter

pchila committed:
Merge remote-tracking branch 'upstream/main' into user-agent-parsing

pchila committed:
make gotidy

pchila committed:
make generate

TylerHelmuth commented:
@pchila needs a changelog then we'll be good to go

pchila committed:
Added changelog

pchila commented:
@TylerHelmuth added the changelog and now all the checks are green 🎉

TylerHelmuth commented:
Looks like your new dependency is conflicting with a renovatebot update that was merged.

pchila committed:
Merge remote-tracking branch 'upstream/main' into user-agent-parsing
